jerry-emscripten-simulator/jerry.c

```diff
@@ -0,0 +1,792 @@
+/* Copyright 2016 Pebble Technology Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerry-api.h"
+
+#include <emscripten/emscripten.h>
+
+#include <string.h>
+
+#define TYPE_ERROR \
+    jerry_create_error(JERRY_ERROR_TYPE, NULL);
+
+#define TYPE_ERROR_ARG \
+    jerry_create_error( \
+      JERRY_ERROR_TYPE, (const jerry_char_t *)"wrong type of argument");
+
+#define TYPE_ERROR_FLAG \
+    jerry_create_error( \
+        JERRY_ERROR_TYPE, \
+        (const jerry_char_t *)"argument cannot have an error flag");
+
+////////////////////////////////////////////////////////////////////////////////
+// Parser and Executor Function
+////////////////////////////////////////////////////////////////////////////////
+
+// Note that `is_strict` is currently unsupported by emscripten
+jerry_value_t jerry_eval(const jerry_char_t *source_p, size_t source_size,
+                         bool is_strict) {
+  // FIXME: use is_strict
+  (void)is_strict;
+
+  return (jerry_value_t)EM_ASM_INT({
+      // jerry_eval() uses an indirect eval() call,
+      // so the global execution context is used.
+      // Also see ECMA 5.1 -- 10.4.2 Entering Eval Code.
+      var indirectEval = eval;
+      try {
+        return __jerryRefs.ref(indirectEval(Module.Pointer_stringify($0, $1)));
+      } catch (e) {
+        var error_ref = __jerryRefs.ref(e);
+        __jerryRefs.setError(error_ref, true);
+        return error_ref;
+      }
+    }, source_p, source_size);
+}
+
+size_t jerry_parse_and_save_snapshot(const jerry_char_t *source_p,
+                                     size_t source_size,
+                                     bool is_for_global, bool is_strict,
+                                     uint8_t *buffer_p, size_t buffer_size) {
+  EM_ASM(throw new Error('Not implemented'));
+  (void)source_p;
+  (void)source_size;
+  (void)is_for_global;
+  (void)is_strict;
+  (void)buffer_p;
+  (void)buffer_size;
+  return 0;
+}
+
+jerry_value_t jerry_parse(const jerry_char_t *source_p, size_t source_size,
+                          bool is_strict) {
+  EM_ASM(throw new Error('Not implemented'));
```

Comment:
The ones that deal with snapshots aren't supported. Supporting those w/o the actual JerryScript engine running would mean we'd have to translate the bytecode back to .js and then run it in the host engine. It's possible I guess, but it's beyond the scope that we are trying to do.

I imagine `jerry_parse` and `jerry_run` could be implemented but we haven't done so just because we didn't have a need for it.